scripts/pc2_interactive.sh

```diff
@@ -108,11 +108,13 @@ KillValidator()
 HandleTerminateFromPC2()
 {
 	DEBUG echo "Received TERMINATE signal from PC2"
+	DEBUG echo "Killing off submission process group $submissionpid and all children"
 	KillValidator
-	if test -n "$submissionpid" -a -d /proc/$submissionpid
+	# Kill off all our kids
+	pkill -9 -P $$
+	if test -n "$submissionpid"
 	then
-		DEBUG echo "Killing off submission process $submissionpid"
-		kill -9 "$submissionpid"
+		kill -9 -"$submissionpid"
```

Comment:
i also don't understand this change of adding a - before the pid?

Reply:
kill with a negative PID kills the process group with that ID.  If we put the submission in it's own session (setsid taskset ...), then we can kill that whole process group when we're done to make sure anyone in that group is killed.  This does NOT catch the case where a forked process does it's own setsid() call, but it will catch those who don't.  The ones that do the setsid() call will be killed when we write a 1 to the cgroup kill file, which we do before we initialize the cgroup.

Reply:
Where is that documented?

Reply:
`man kill`
```
DESCRIPTION
       The default signal for kill is TERM.  Use -l or -L to list available signals.  Particularly useful signals include HUP, INT, KILL, STOP, CONT, and 0.
       Alternate signals may be specified in three ways: -9, -SIGKILL or -KILL.  **Negative PID values may be used to choose whole  process  groups**;  see  the
       PGID column in ps command output.  A PID of -1 is special; it indicates all processes except the kill process itself and init.
```

Reply:
hmm,  you are referencing kill though which is bash internal.    
```
       NOTES  Your shell (command line interpreter) may have a  built-in  kill
              command.   You  may  need  to  run the command described here as
              /bin/kill to solve the conflict.
```

Reply:
Oh.  Excellent point.